packages/state-transition/src/cache/epochCache.ts

```diff
@@ -500,40 +572,88 @@ export class EpochCache {
 
-    // Roll current proposers into previous proposers for metrics
+    // move current to previous
+    this.previousShuffling = this.currentShuffling;
+    this.previousDecisionRoot = this.currentDecisionRoot;
     this.proposersPrevEpoch = this.proposers;
 
-    const currentProposerSeed = getSeed(state, this.currentShuffling.epoch, DOMAIN_BEACON_PROPOSER);
-    this.proposers = computeProposers(currentProposerSeed, this.currentShuffling, this.effectiveBalanceIncrements);
+    // move next to current or calculate upcoming
+    this.currentDecisionRoot = this.nextDecisionRoot;
+    if (this.nextShuffling) {
+      // was already pulled from the ShufflingCache to the EpochCache (should be in most cases)
+      this.currentShuffling = this.nextShuffling;
+    } else {
+      this.shufflingCache?.metrics?.shufflingCache.nextShufflingNotOnEpochCache.inc();
+      this.currentShuffling =
+        this.shufflingCache?.getSync(upcomingEpoch, this.nextDecisionRoot, {
+          state,
+          // have to use the "nextActiveIndices" that were saved in the last transition here to calculate
+          // the upcoming shuffling if it is not already built (similar condition to the below computation)
+          activeIndices: this.nextActiveIndices,
+        }) ??
+        // allow for this case during testing where the ShufflingCache is not present, may affect perf testing
+        // so should be taken into account when structuring tests.  Should not affect unit or other tests though
+        computeEpochShuffling(state, this.nextActiveIndices, upcomingEpoch);
+    }
+    const upcomingProposerSeed = getSeed(state, upcomingEpoch, DOMAIN_BEACON_PROPOSER);
+    // next epoch was moved to current epoch so use current here
+    this.proposers = computeProposers(upcomingProposerSeed, this.currentShuffling, this.effectiveBalanceIncrements);
+
+    // handle next values
+    this.nextDecisionRoot = epochTransitionCache.nextShufflingDecisionRoot;
+    this.nextActiveIndices = epochTransitionCache.nextShufflingActiveIndices;
+    if (this.shufflingCache) {
+      this.nextShuffling = null;
+      // This promise will resolve immediately after the synchronous code of the state-transition runs. Until
+      // the build is done on a worker thread it will be calculated immediately after the epoch transition
+      // completes.  Once the work is done concurrently it should be ready by time this get runs so the promise
+      // will resolve directly on the next spin of the event loop because the epoch transition and shuffling take
+      // about the same time to calculate so theoretically its ready now.  Do not await here though in case it
+      // is not ready yet as the transition must not be asynchronous.
+      this.shufflingCache
+        .get(epochAfterUpcoming, this.nextDecisionRoot)
+        .then((shuffling) => {
+          if (!shuffling) {
+            throw new Error("EpochShuffling not returned from get in afterProcessEpoch");
```

Comment:
add epoch, decision root to the error message

Reply:
It will be added in the `.catch`.  Realistically that condition should never happen but wanted to check for it to make sure just in case so throwing the error message there to add additional context to the log in the catch.